### DISCUSSION TITLE
What is the policy for contributions using java based tests instead of groovy?

### DISCUSSION THREAD

**Discussion by rferreira** [micronaut-projects/micronaut-core#5914] [Q&A] [answered]
Hello folks 

I'm a small code contributor for micronaut (mostly views) and I'm planning a largish contribution and I feel pretty confident that I would be able to write more robust testing if I could write them in good old java/junit but all testing appears to be done in groovy/spock which seems like a fine combination but not what I'm most productive with. It's not a big deal if the answer is "no" and all tests must be written in groovy/spock, I'm just looking for confirmation. 

Thanks!

**Comment by graemerocher** [comment, marked as answer]
We have several test with JUnit 5 and Java as well, no problem using it

**Reply by graemerocher** [reply]
To clarify we probably wouldn't accept JUnit 4 or TestNG but JUnit 5 and spock share the same runtime so that is ok

**Reply by ilopmar** [reply]
I don't agree with this. I think if we are testing almost everything with Spock, we should keep it that way. Specially if a module only uses Spock.
Not all of us are "fluent" with JUnit5 (I'm personally don't like writing tests in Java/JUnit5 and every time I write a guide and I need to write JUnit5 tests I "cry inside") and once we merge JUnit5 tests, then we need to maintain them forever.

So -1 for me on this. If the module has only Spock tests, then that's what it should be for new contributions.

**Reply by sdelamo** [reply]
I would prefer Spock tests as well. @rferreira However, if you want to contribute them initially as JUnit 5 in the PR then I can help you to translate them to Spock.

**Reply by graemerocher** [reply]
@ilopmar @sdelamo I wasn't referring to modules, I was referring to `micronaut-core` which already has a mix of spock and JUnit 5 tests so I don't see any harm in additional tests and FWIW IMO if we take this stance we are just limiting contributions since vastly more contributors know JUnit/Java than they do Spock/Groovy.

**Reply by ilopmar** [reply]
He was explicitly asking about micronaut-views. 

Even if we have JUnit5 tests in core, I think we should have more consistent with the tests. What we should have probably done is what Sergio proposes. If we received a PR with JUnit5 we help the user to convert them to Spock (or we do it before we merge the PR).
If we decided to use Spock it was for some reasons: easy to read/write tests, more powerful, full use of Groovy for testing,... so we should stick with it. If not, we should have chosen JUnit (no please!)

**Reply by graemerocher** [reply]
I standby my point that we are just limiting contributions. If we are going to have a policy of rejecting PRs for being in JUnit 5 I strongly disagree with that and believe we will have to take the burden on of porting the tests to Spock for the user and not push back and ask them to learn a new language just to write tests. People shouldn't be expected to learn Groovy because of silly personal preferences.

**Reply by ilopmar** [reply]
As I said, yes, we should have ported the tests to Spock in the same PR before merging it.
I don't think using Spock is a _silly personal preference_ but a choice we (in this case you) made when starting the framework (and the right one from my point of view). The same as using Gradle instead of Maven.

**Reply by graemerocher** [reply]
Absolutely, I love Spock and think it is the best 🙂

At the same time I appreciate that Java is 70% of our user base and Groovy is single digits and the Micronaut codebase is 99% written in Java (excluding tests). I am also empathetic now to folks who have not used Groovy or Kotlin.

Also we are not talking about some monstrosity here, JUnit 5 is a vast improvement to JUnit 4 and actually quite pleasant in many ways. 

In my experience as well there are cases where you simply have to use Java to write tests for a Java API because Groovy doesn't allow you to test how a Java API works in terms of generics, lambdas etc. 

It will also not be possible to effectively write native image tests with Spock with the upcoming GraalVM support so I anticipate there will be more and more JUnit tests around in modules.

In addition if a subproject has correct documentation then it must have a mix of Spock and JUnit 5 anyway to support multi language documentation.

So I don't agree with a blanket ban on JUnit 5 tests and stand by that.

**Reply by jameskleeh** [reply]
We should never reject a PR because it contains Junit5 tests. I think it's up to the person who has taken primary responsibility of that module how they want to proceed. To me there are 2 choices, either merge the PR as is, or take the task on to convert them for the contributor.

From my perspective if the PR has to include dependencies to Junit to get the tests to pass I would rather switch the tests to Spock. If the module already has Junit tests then I likely wouldn't bother.

We shouldn't put the responsibility of doing the conversion on the submitter in order to get the PR merged in any case.
